File: 083_Average of Levels in Binary Tree.py
"""
Average of Levels in Binary Tree
Calculate average value of nodes on each level.
"""
from typing import Optional, List

class TreeNode:
    def __init__(self,val=0,left=None,right=None):
        self.val=val; self.left=left; self.right=right

def average_of_levels(root: Optional[TreeNode]) -> List[float]:
    if not root: return []
    from collections import deque
    q = deque([root])
    res = []
    while q:
        n = len(q); s=0
        for _ in range(n):
            node = q.popleft(); s += node.val
            if node.left: q.append(node.left)
            if node.right: q.append(node.right)
        res.append(s/n)
    return res

if __name__ == '__main__':
    root = TreeNode(3)
    root.left = TreeNode(9)
    root.right = TreeNode(20)
    root.right.left = TreeNode(15)
    root.right.right = TreeNode(7)
    print('Averages:', average_of_levels(root))

# Time Complexity: O(n)
# Space Complexity: O(n)

File: 083_Average of Levels in Binary Tree.cpp
// Average of Levels in Binary Tree
// Compute average value for nodes on each level.
#include <bits/stdc++.h>
using namespace std;
struct TreeNode{int val; TreeNode* left; TreeNode* right; TreeNode(int x): val(x), left(nullptr), right(nullptr) {}};

vector<double> averageOfLevels(TreeNode* root){
    vector<double> res;
    if(!root) return res;
    queue<TreeNode*> q; q.push(root);
    while(!q.empty()){
        long long sum=0; int n=q.size();
        for(int i=0;i<n;++i){ TreeNode* t=q.front(); q.pop(); sum+=t->val; if(t->left) q.push(t->left); if(t->right) q.push(t->right); }
        res.push_back((double)sum/n);
    }
    return res;
}

int main(){
    TreeNode* root = new TreeNode(3);
    root->left = new TreeNode(9);
    root->right = new TreeNode(20);
    root->right->left = new TreeNode(15);
    root->right->right = new TreeNode(7);
    auto v = averageOfLevels(root);
    cout<<"Averages: "; for(double d:v) cout<<d<<" "; cout<<"\n";
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(n)
